### DISCUSSION TITLE
Deploy the Semaphore protocol on zkSync

### DISCUSSION THREAD

**Discussion by vplasencia** [zkSync-Community-Hub/zksync-developers#365] [Compilers]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

0.8.4 / 0.8.23

### zksolc/zkvyper version?

v1.3.23

### Explain the issue

Hello 👋

I'm trying to deploy the [Semaphore protocol](https://github.com/semaphore-protocol/semaphore) on zkSync. I'm trying with Semaphore v3 which is the latest stable version and also Semaphore v4 which is the new version.

I have some questions:

1. How can I deploy bytecode?

To deploy Semaphore v3, it is necessary to get the Poseidon code from a JavaScript library: https://github.com/semaphore-protocol/semaphore/blob/main/packages/hardhat/src/tasks/deploy-semaphore.ts#L59-L73

I'm trying to deploy Poseidon and I get the following error: `Error: The bytecode length in bytes must be divisible by 32!`

Here is the code of the script: https://github.com/vplasencia/semaphore-zksync/blob/main/semaphorev3/deploy/poseidon/deploy-poseidon.ts

More info:

This is the Hardhat task to deploy Semaphore v3: https://github.com/semaphore-protocol/semaphore/blob/main/packages/hardhat/src/tasks/deploy-semaphore.ts

2. How can I link libraries?

After deploying the Poseidon bytecode, it is necessary to link the library, how can I recreate this code: https://github.com/semaphore-protocol/semaphore/blob/main/packages/hardhat/src/tasks/deploy-semaphore.ts#L76-L80

If the code of the library is already there in the `.sol` file, I can deploy the library using `yarn hardhat deploy-zksync:libraries --private-key <PRIVATE_KEY>` and everything should be fine when I want to compile. But what would be the flow when not all the code is in the `.sol` file? 

3. When I try to deploy this Poseidon contract, it does not finish the deployment. Any idea why this is happening?

https://github.com/vimwitch/poseidon-solidity/blob/main/contracts/PoseidonT3.sol

### Link to your repo?

https://github.com/vplasencia/semaphore-zksync

**Comment by uF4No** [comment]
Hi @vplasencia 

> How can I deploy bytecode?

zkSync is not EVM bytecode compatible, but source-code compatible. This means that you can have the same source code that you'd use in other EVM protocols, but zkSync has its own compilers (zksolc and zkvyper), that generate a different bytecode compatible with zkSync's ZKEVM. You'd need to get the source code of the contract, compile it with `zksolc` to generate the bytecode (which is done automatically if the `hardhat-zksync-solc` plugin is installed and added in the `hardhat.config.ts` file), and then deploy it.

> How can I link libraries?

Not sure if I understand this correctly. You could deploy the library first and then reference it by its address in a separate contract.

> When I try to deploy this Poseidon contract, it does not finish the deployment. Any idea why this is happening?

Are you getting any errors?

**Comment by vplasencia** [comment]
Hey @uF4No! Thank you so much for the answers.

1. Makes sense. 

2. It was related to deploying the bytecode and then linking the library. For libraries in general I saw that I can deploy them first and in the `hardhat.config.ts` file, in the `settings` property inside the `zksolc` property it will add the library addresses. 

3. It's like it's frozen. I tried again, it takes about 1 or 2 hours compiling, and then it throws the following error:

```console
vivian@mbp-de-vivian my-zksync-project % npm run compile

> compile
> hardhat compile

Compiling contracts for zkSync Era with zksolc v1.4.0 and solc v0.8.17
Compiling 1 Solidity file
An unexpected error occurred:

Error: Command failed: /Users/vivian/Library/Caches/hardhat-nodejs/compilers-v2/zksolc/zksolc-v1.4.0 --standard-json   --solc /Users/vivian/Library/Caches/hardhat-nodejs/compilers-v2/macosx-amd64/solc-macosx-amd64-v0.8.17+commit.8df45f5f --detect-missing-libraries
/Users/vivian/Library/Caches/hardhat-nodejs/compilers-v2/macosx-amd64/solc-macosx-amd64-v0.8.17+commit.8df45f5f error: 

    at ChildProcess.exithandler (node:child_process:421:12)
    at ChildProcess.emit (node:events:514:28)
    at ChildProcess.emit (node:domain:489:12)
    at maybeClose (node:internal/child_process:1105:16)
    at Socket.<anonymous> (node:internal/child_process:457:11)
    at Socket.emit (node:events:514:28)
    at Socket.emit (node:domain:489:12)
    at Pipe.<anonymous> (node:net:337:12) {
  code: 1,
  killed: false,
  signal: null,
  cmd: '/Users/vivian/Library/Caches/hardhat-nodejs/compilers-v2/zksolc/zksolc-v1.4.0 --standard-json   --solc /Users/vivian/Library/Caches/hardhat-nodejs/compilers-v2/macosx-amd64/solc-macosx-amd64-v0.8.17+commit.8df45f5f --detect-missing-libraries'
}
```

**Reply by uF4No** [reply]
Looks like it's an issue with `solc`. We've reported it [here](https://github.com/ethereum/solidity/issues/14885).

Please try compiling with the `forceEvmla: true` in the zksolc settings. You can [find more info in the compiler docs](https://docs.zksync.io/build/tooling/hardhat/hardhat-zksync-solc.html#configuration)

**Reply by uF4No** [reply]
@vplasencia let us know if you managed to compile the contract with `forceEvmla: true`

**Reply by vplasencia** [reply]
Hey @uF4No!

Thank you for reporting the issue. I compiled the `PoseidonT3.sol` contract with `forceEvmla: true` in the zksolc settings and it works fine. 

One question, when I compile this contract [Feedback.sol](https://github.com/vplasencia/semaphore-zksync/blob/main/semaphorev4/contracts/Feedback.sol), I think that it should generate the artifacts for all the contracts it's using so that when I want to deploy that contract, all the artifacts are there, but it does not generate all the artifacts. Should I add another config for that? 

I think if I copy/paste all the contracts that the `Feedback.sol`contract uses, everything should work fine.

**Reply by uF4No** [reply]
Glad you were able to compile PoseidonT3.sol with `forceEvmla: true` 🎉

Regarding the `Feedback.sol` contract, I don't think you need any other config. I don't see that `Feedback.sol` imports any other contracts, just the `ISemaphore` interface so it it would not generate any other artifacts. Try to deploy it locally or to the zkSync Sepolia testnet to confirm it works as expected.

**Reply by vplasencia** [reply]
Hey! Regarding the artifacts, I get what you mean, but when I compile that contract in the Hardhat environment without `zksolc`, it generates all the artifacts that are needed to deploy. If I compile this contract with `zksolc`, it only generates the `ISemaphore.sol` artifacts and we need more artifacts to deploy that contract.

If I copy/paste manually all the contracts that Semaphore is using and compile them all together, it works. I was able to deploy all of them successfully using `forceEvmla: true`. However, developers would need to copy/paste all the contracts that Semaphore is using to make a contract work.

I was able to interact with the contract on `zkSync Era Sepolia Testnet`, but the transactions to verify Semaphore proofs (the zk part in contracts) were reverting with this error: `shortMessage: 'execution reverted (unknown custom error)'`. I will try other ideas to check if the error is related to something in the code or if the Verifier is not working properly on zkSync and will let you know. 

Thank you for all the help.

**Reply by uF4No** [reply]
hey @vplasencia sorry, I'm not familiar with Semaphore so I'm don't really understand what other artifacts are required for the `Feedback.sol` contract to work as it does not import any contracts, just the `ISemaphore.sol` interface. Can you put all the contracts needed inside the same project and compile them with zksolc? that should generate all the artifacts you need.

Regarding the workaround you've done and the failed transactions, you can [debug them by replaying them locally using era-test-node](https://docs.zksync.io/build/test-and-debug/era-test-node.html#replay-remote-transactions-locally).

**Comment by longbipubacr** [comment]
It seems like you're encountering several issues related to deploying the Semaphore protocol on zkSync, including deploying bytecode, linking libraries, and deployment failures with the Poseidon contract. To address these:

For deploying bytecode, ensure that the bytecode length is divisible by 32 to meet zkSync's requirements.

To link libraries, you need to use the link method provided by Hardhat, passing the library's address and position in the bytecode.

If not all the code is in the .sol file, you may need to deploy the library separately and then link it during contract deployment.

The deployment failure with the Poseidon contract could be due to various reasons, such as gas limit issues, incorrect contract initialization, or compatibility issues with zkSync. Ensure thorough testing and debugging to identify and resolve the specific cause.

Continuously refer to documentation, community resources, and debugging tools to troubleshoot and resolve these issues effectively.

**Comment by fitzgibbonsdustin** [comment]
1. To deploy bytecode, ensure that the bytecode length is divisible by 32. You may need to adjust the bytecode or padding to meet this requirement.

2. To link libraries, you can use the Hardhat task deploy-zksync:libraries with the private key. If the library code is not all in the .sol file, ensure that the necessary code is available and accessible during deployment.

3. If the deployment of the Poseidon contract is not completing, double-check the deployment process, ensure that all required parameters are provided correctly, and debug any potential issues with the contract code or deployment environment.

**Comment by duongnd-183902** [comment]
Hi @vplasencia
Have you deployed Poseidon hash yet?

**Reply by duongnd-183902** [reply]
I would like to emphasize that I am able to deploy the PoseidonT3 contract by configuring forceEvmla: true. However, I've come across a cautionary note indicating that enabling the forceEvmla field may lead to the following adverse effects:

Lack of support for recursion.
Lack of support for internal function pointers.
Potential impact on contract size and performance.
For Solidity versions older than 0.8, only this compilation mode is available, and it is used by default.
Are there any other documents that provide further clarification on this issue? Additionally, should I consider using this configuration forceEvmla: true for other contracts?

**Reply by bxpana** [reply]
@duongnd-183902 if you haven't already, let's go ahead and open a separate discussion post for this.

**Comment by yamyv** [comment]
Issue: Deploying Semaphore protocol on zkSync, encountering errors while deploying Poseidon contract and linking libraries.

1. Error deploying Poseidon contract: The bytecode length in bytes must be divisible by 32!
2. Question about linking libraries when not all code is in the .sol file.
3. Deployment of Poseidon contract does not finish.

Solution:
1. Ensure that the bytecode length of the Poseidon contract is divisible by 32.
2. When not all code is in the .sol file, deploy the library separately and then link it during contract deployment.
3. Investigate possible reasons for the deployment of Poseidon contract not finishing, such as gas limits or contract code issues.
 Note: Further debugging may be required based on specific error messages or logs encountered during deployment.

**Comment by shipefred** [comment]
Which compiler are you using: zksolc
Version of Solidity: 0.8.4 / 0.8.23
zksolc version: v1.3.23
Issue Explanation: Seeking assistance with deploying Semaphore protocol on zkSync, specifically Semaphore v3 and v4. Questions include deploying bytecode, linking libraries, and deployment errors.
Repo Link: https://github.com/vplasencia/semaphore-zksync
